src/chain.h

```diff
@@ -429,11 +429,14 @@ class CChain
 private:
     std::vector<CBlockIndex *> vChain;
 
+    // hold a copy of the tip outside of the vector so it can be accessed without holding cs_main
+    std::atomic<CBlockIndex *> tip;
+
 public:
     /** Returns the index entry for the genesis block of this chain, or NULL if none. */
     CBlockIndex *Genesis() const { return vChain.size() > 0 ? vChain[0] : NULL; }
     /** Returns the index entry for the tip of this chain, or NULL if none. */
-    CBlockIndex *Tip() const { return vChain.size() > 0 ? vChain[vChain.size() - 1] : NULL; }
+    CBlockIndex *Tip() const { return tip; }
```

Comment:
This does not return null anymore

-/** Returns the index entry for the tip of this chain, or NULL if none. */
+/** Returns the index entry for the tip of this chain */

Reply:
tip can be nullptr in the same degenerate cases that Tip() returned null